sdk/schemaregistry/schema-registry-avro/src/schemaRegistryAvroSerializer.ts

```diff
@@ -0,0 +1,226 @@
+  /** Schema ID */
+  id: string;
+
+  /** avsc-specific representation for schema */
+  type: avro.Type;
+}
+
+/**
+ * Options for Schema
+ */
+export interface SchemaRegistryAvroSerializerOptions {
+  /**
+   * When true, register new schemas passed to serialize. Otherwise, and by
+   * default, fail if schema has not already been registered.
+   *
+   * Automatic schema registration is NOT recommended for production scenarios.
+   */
+  autoRegisterSchemas?: boolean;
+}
+
+/**
+ * Avro serializer that obtains schemas from a schema registry and does not
+ * pack schemas into its payloads.
+ */
+export class SchemaRegistryAvroSerializer {
+  /**
+   * Creates a new serializer.
+   *
+   * @param registry Schema Registry where schemas are registered and obtained.
+   *                 Usually this is a SchemaRegistryClient instance.
+   *
+   * @param schemaGroup The schema group to use when making requests to the
+   *                    registry.
+   */
+  constructor(
+    registry: SchemaRegistry,
+    schemaGroup: string,
+    options?: SchemaRegistryAvroSerializerOptions
+  ) {
+    this.registry = registry;
+    this.schemaGroup = schemaGroup;
+    this.autoRegisterSchemas = options?.autoRegisterSchemas ?? false;
+  }
+
+  /** The schema group to use when making requests to the registry*/
+  public readonly schemaGroup: string;
+
+  /**
+   * Schema Registry where schemas are registered and obtained.
+   * Usually this is a SchemaRegistryClient instance.
+   */
+  public readonly registry: SchemaRegistry;
+
+  /**
+   * When true, register new schemas passed to serialize. Otherwise, and by
+   * default, fail if schema has not already been registered.
+   *
+   * Automatic schema registration is NOT recommended for production scenarios.
+   */
+  public autoRegisterSchemas: boolean;
+
+  // REVIEW: signature.
+  //
+  // - Better to serialize into a stream? I aborted that for now as I wanted to
+  //   do the simplest thing that could possibly work first to make sure there
+  //   were no blockers in our dependencies. I also wanted to get feedback on
+  //   what the API shape should be before diving into that.
+  //
+  // - This type should ultimately be able to implement a core ObjectSerializer
+  //   interface. Do we know what that would look like? Maybe it takes `any` as
+  //   the format-specific schema/type info? Or does it always take a
+  //   format-specific schema string?
+  //
+  //   The C#/Java approach of passing Type and assuming every serializer can
+  //   get its schema by reflecting on the type does not work for JavaScript. We
+  //   need to support arbitrary objects that match a schema.
+  //
+  //   Maybe each format expects a different property on this arg so that you
+  //   could at least pass enough info for multiple formats, and then your call
+  //   to ObjectSerializer is at least not tied to a single format?
+  //
+  // - Should we wrap all errors thrown by avsc to avoid having our exception //
```

Comment:
Are errors from avsc well documented? It sounds nice if you can do it, but I guess it depends too on if it's something the user would want to handle programmatically.

For example, if avsc is going to yell at the developer while they're coding and provide helpful messages, I don't know if much wrapping is required, but if it's throwing runtime exceptions they should handle in production, I'd really want us to own that contract.

Reply:
100% agree on that distinction, but I don't have the data right now. Unfortunately, the errors are not documented: https://github.com/mtth/avsc/wiki/API#typetobufferval

I don't expect errors to be of the kind that you would want to handle them in production, but it's possible that I'm missing data. I've got a note to keep thinking about this in #10950 and I think it'll be OK to leave this undone for beta 1. Would you agree?